File: content/snippets/python/interview_lcci_04_02_minimum_height_tree.py
class Node:
    def __init__(self, val: int) -> None:
        self.val = val
        self.left: Node | None = None
        self.right: Node | None = None


def build(nums: list[int], lo: int, hi: int) -> Node | None:
    if lo > hi:
        return None
    mid = (lo + hi) // 2
    root = Node(nums[mid])
    root.left = build(nums, lo, mid - 1)
    root.right = build(nums, mid + 1, hi)
    return root


def inorder(root: Node | None) -> list[int]:
    if root is None:
        return []
    return inorder(root.left) + [root.val] + inorder(root.right)


def main() -> None:
    nums = [1, 2, 3, 4, 5, 6, 7]
    print(" ".join(map(str, inorder(build(nums, 0, len(nums) - 1)))))


if __name__ == "__main__":
    main()

File: content/snippets/typescript/interview_lcci_04_02_minimum_height_tree.ts
class Node {
  constructor(
    public val: number,
    public left: Node | null = null,
    public right: Node | null = null
  ) {}
}

function build(a: number[], lo: number, hi: number): Node | null {
  if (lo > hi) return null;
  const mid = Math.floor((lo + hi) / 2);
  return new Node(a[mid], build(a, lo, mid - 1), build(a, mid + 1, hi));
}

function inorder(n: Node | null, out: number[]): void {
  if (!n) return;
  inorder(n.left, out);
  out.push(n.val);
  inorder(n.right, out);
}

const nums = [1, 2, 3, 4, 5, 6, 7];
const root = build(nums, 0, nums.length - 1);
const out: number[] = [];
inorder(root, out);
console.log(out.join(" "));
